titus-common-api/src/main/java/com/netflix/titus/api/model/PaginationEvaluator.java

```diff
@@ -0,0 +1,200 @@
+        int offset = getOffset(itemsCopy, cursorId, cursorTimestamp);
+
+        int totalItems = itemsCopy.size();
+        boolean isEmptyResult = offset >= totalItems;
+        boolean hasMore = totalItems > (offset + page.getPageSize());
+        int endOffset = Math.min(totalItems, offset + page.getPageSize());
+        int cursorPosition = endOffset - 1;
+        int numberOfPages = numberOfPages(page, totalItems);
+        int pageNumber = Math.min(numberOfPages, offset / page.getPageSize());
+
+        Pagination pagination = new Pagination(
+                page.toBuilder().withPageNumber(pageNumber).build(),
+                hasMore,
+                numberOfPages,
+                totalItems,
+                totalItems == 0 ? "" : encode(itemsCopy.get(cursorPosition)),
+                totalItems == 0 ? 0 : cursorPosition
+        );
+
+        List<T> pageItems = isEmptyResult ? Collections.emptyList() : itemsCopy.subList(offset, endOffset);
+        return PageResult.pageOf(pageItems, pagination);
+    }
+
+    /**
+     * Offset is a position of an element after the one pointed by the cursor. If cursor does not point to any
+     * existing element, the offset points to the smallest element larger than cursor.
+     */
+    private int getOffset(List<T> itemsCopy, String cursorId, long cursorTimestamp) {
+        if (itemsCopy.isEmpty()) {
+            return 0;
+        }
+
+        Function<T, Integer> comparator = item -> {
+            long itemTimestamp = Evaluators.getOrDefault(timestampExtractor.apply(item), 0L);
+            int timestampCmp = Long.compare(cursorTimestamp, itemTimestamp);
+            if (timestampCmp != 0) {
+                return timestampCmp;
+            }
+
+            String itemId = idExtractor.apply(item);
+            return cursorId.compareTo(itemId);
+        };
+
+        int pos = CollectionsExt.binarySearchLeftMost(itemsCopy, comparator);
+        return pos >= 0 ? (pos + 1) : -(pos + 1);
+    }
+
+    @VisibleForTesting
+    String encode(T value) {
+        String id = idExtractor.apply(value);
+        long timeStamp = Evaluators.getOrDefault(timestampExtractor.apply(value), 0L);
+        return Base64.getEncoder().encodeToString((id + '@' + timeStamp).getBytes());
+    }
+
+    private static Pair<String, Long> decode(String encoded) {
+        String decoded = new String(Base64.getDecoder().decode(encoded.getBytes()));
+
+        Matcher matcher = CURSOR_RE.matcher(decoded);
+        if (!matcher.matches()) {
+            throw new IllegalArgumentException(
+                    String.format("Not a valid cursor value: encoded=%s, decoded=%s", encoded, decoded)
+            );
+        }
+
+        String jobId = matcher.group(1);
+        long timestamp = Long.parseLong(matcher.group(2));
+
+        return Pair.of(jobId, timestamp);
+    }
+
+    private static int numberOfPages(Page page, int totalItems) {
+        return (totalItems + page.getPageSize() - 1) / page.getPageSize();
```

Comment:
Do we validate somewhere that the user hasn't set pageSize to 0?

Reply:
Also, I'm a little confused about the calculation here. If totalItems is 10 and pageSize is 4, I would expect there to be 3 pages. Seems like this would return 4.

Reply:
No, we do not do that. I will add this check.
(10 + 4  - 1)/4 = 3. The calculation is correct.

Reply:
I will support page size 0 as a special case, which returns empty list. This is useful if a client is interested only in the pagination data (mostly total items).

Reply:
Yeah, I must have misread it before. Looks correct.